android/expoview/src/main/java/host/exp/exponent/ExponentManifest.kt

```diff
@@ -0,0 +1,656 @@
+      val url = URL(iconUrl)
+      val connection = url.openConnection() as HttpURLConnection
+      connection.doInput = true
+      connection.connect()
+      val input = connection.inputStream
+      val bitmap = BitmapFactory.decodeStream(input)
+      val width = bitmap.width
+      val height = bitmap.height
+      if (width <= MAX_BITMAP_SIZE && height <= MAX_BITMAP_SIZE) {
+        memoryCache.put(iconUrl, bitmap)
+        return bitmap
+      }
+      val maxDimension = max(width, height)
+      val scaledWidth = width.toFloat() * MAX_BITMAP_SIZE / maxDimension
+      val scaledHeight = height.toFloat() * MAX_BITMAP_SIZE / maxDimension
+      val scaledBitmap =
+        Bitmap.createScaledBitmap(bitmap, scaledWidth.toInt(), scaledHeight.toInt(), true)
+      memoryCache.put(iconUrl, scaledBitmap)
+      scaledBitmap
+    } catch (e: IOException) {
+      EXL.e(TAG, e)
+      BitmapFactory.decodeResource(context.resources, R.mipmap.ic_launcher)
+    } catch (e: Throwable) {
+      EXL.e(TAG, e)
+      BitmapFactory.decodeResource(context.resources, R.mipmap.ic_launcher)
+    }
+  }
+
+  fun getColorFromManifest(manifest: RawManifest): Int {
+    val colorString = manifest.getPrimaryColor()
+    return if (colorString != null && ColorParser.isValid(colorString)) {
+      Color.parseColor(colorString)
+    } else {
+      R.color.colorPrimary
+    }
+  }
+
+  fun isAnonymousExperience(manifest: RawManifest): Boolean {
+    return try {
+      val id = manifest.getLegacyID()
+      id.startsWith(ANONYMOUS_EXPERIENCE_PREFIX)
+    } catch (e: JSONException) {
+      false
+    }
+  }
+
+  private val localKernelManifest: RawManifest
+    get() = try {
+      val manifest = JSONObject(ExponentBuildConstants.BUILD_MACHINE_KERNEL_MANIFEST)
+      manifest.put(MANIFEST_IS_VERIFIED_KEY, true)
+      ManifestFactory.getRawManifestFromJson(manifest)
+    } catch (e: JSONException) {
+      throw RuntimeException("Can't get local manifest: $e")
+    }
+
+  private val remoteKernelManifest: RawManifest?
+    get() = try {
+      val inputStream = context.assets.open(EMBEDDED_KERNEL_MANIFEST_ASSET)
+      val jsonString = IOUtils.toString(inputStream)
+      val manifest = JSONObject(jsonString)
+      manifest.put(MANIFEST_IS_VERIFIED_KEY, true)
+      ManifestFactory.getRawManifestFromJson(manifest)
+    } catch (e: Exception) {
+      KernelProvider.instance.handleError(e)
+      null
+    }
+
+  val kernelManifest: RawManifest
+    get() {
+      val manifest: RawManifest?
+      val log: String
+      if (exponentSharedPreferences.shouldUseInternetKernel()) {
+        log = "Using remote Expo kernel manifest"
+        manifest = remoteKernelManifest
+      } else {
+        log = "Using local Expo kernel manifest"
+        manifest = localKernelManifest
+      }
+      if (!hasShownKernelManifestLog) {
+        hasShownKernelManifestLog = true
+        EXL.d(TAG, log + ": " + manifest.toString())
+      }
+      return manifest!!
+    }
```

Comment:
I'm nervous to make these properties instead of methods because they perform I/O and can throw exceptions. These getters do so much work that it is helpful to emphasize the "get" part by making them methods instead of properties.